### ISSUE TITLE
Benchmark MergeWatch on withmartian/code-review-benchmark: run, publish, and submit upstream

### ISSUE THREAD

**Issue by santthosh** [mergewatch/mergewatch.ai#314]
Run MergeWatch against [withmartian/code-review-benchmark](https://github.com/withmartian/code-review-benchmark), publish the results, and submit them upstream so MergeWatch appears on the [leaderboard](https://codereview.withmartian.com) alongside CodeRabbit, Greptile, Cursor Bugbot, Qodo, Copilot and the rest.

Closes the ask in #242. The requester's actual words were *"it's hard to know if evaluating this tool will be worth my time without some sort of evidence"* — the deliverable is evidence a stranger can check, not a number we like.

---

## What the benchmark actually is

**Offline set:** 50 PRs across 5 projects and 5 languages — Sentry (Python), Grafana (Go), Cal.com (TypeScript), Discourse (Ruby), Keycloak (Java). Ground truth is human-verified "golden comments" with severity labels (Low / Medium / High / Critical).

**Online set:** continuously sampled fresh PRs from GitHub Archive via BigQuery, matched against the developer's post-review fixes. Recency is the point — it resists training-data memorization.

**Metrics:** precision (share of our comments that were accurate/useful) and recall (share of real issues we caught).

**How a tool is evaluated:** there is no adapter to write. You fork the 50 benchmark PRs into a GitHub org, install your tool there, let it review, add the tool name to the download config, and the pipeline scrapes the comments it left:

```bash
cd offline && uv sync && cp .env.example .env   # GitHub token + LLM API key
uv run python -m code_review_benchmark.step1_download_prs --output results/benchmark_data.json
uv run python -m code_review_benchmark.step2_extract_comments
uv run python -m code_review_benchmark.step2_5_dedup_candidates
uv run python -m code_review_benchmark.step3_judge_comments --dedup-groups results/$(MARTIAN_MODEL)/dedup_groups.json
open analysis/benchmark_dashboard.html
```

Because it reads our real PR comments, **we are benchmarking the shipped product**, not a special mode. That is good for credibility and it means our defaults are part of what is measured — see the configuration section, which is the part most likely to produce a misleading number.

---

## Our defaults will depress recall, and that is a deliberate product choice

This is the single most important thing to get right, because the naive run will make us look worse than we are *at the thing we optimized for* and we should not quietly tune our way out of it.

Roughly a year of work — W1–W11, FP-A through FP-K — has gone into **suppressing** findings: the triage-aware convergence guard, cross-agent dedup, claim-aware grounding, linter deference, abstraction-aware verification, dispute-aware score softening. A recall-weighted benchmark rewards exactly the noisy behavior that work removed.

Defaults that will bite (`packages/core/src/config/defaults.ts`):

| Setting | Default | Effect on the benchmark |
|---|---|---|
| `maxFindings` | 25 | Caps reported issues per PR |
| `maxFiles` | 50 | PRs above this are **skipped entirely** → zero recall on those |
| `skipDrafts` | true | Any forked PR left as draft is skipped |
| `maxTokensPerAgent` | 4096 | Truncation risk on large diffs |
| `codebaseAwareness` | off by default | Out-of-diff context unavailable; some golden comments need it |
| `shouldSkipPR` | docs/lockfile detection | Silently drops qualifying PRs |

Skips are the dangerous one: a skipped PR contributes no comments, so it scores as total recall failure rather than as "correctly declined to review."

**Proposal — publish two arms, not one:**

- **Arm A — shipped defaults.** What a user actually gets. The honest headline number.
- **Arm B — recall-favoring config** (`maxFindings` raised, `codebaseAwareness: true`, skip rules relaxed, `maxFiles` lifted). Shows the precision/recall tradeoff curve rather than a single point.

Publishing both is more persuasive than either alone: it demonstrates the suppression is a tuned choice, not a capability ceiling. Submit Arm A upstream as the official entry.

- [ ] Decide and record the exact config for each arm before any run
- [ ] Verify no forked repo carries an upstream `.mergewatch.yml` that silently changes behavior (the fork inherits whatever the source repo had)

---

## Model pinning and disclosure

Results are meaningless without stating which model produced them. We are a harness over frontier models, so the model is a variable we control and must disclose.

- **Pin explicitly** via `infra/params/*.env`, currently `us.anthropic.claude-opus-4-6-v1`.
- **Record** model id, `mergewatch.ai` commit SHA, config file, and run date alongside every result.
- Sonnet 5 is the intended destination (#262) but is blocked by an `AccessDeniedException` on our Bedrock account. If it unblocks before the run, decide deliberately whether to benchmark on it — do not let the deployed default drift mid-run.

**Note for whoever writes the results up:** #264 found prod and dev silently running different models for months. A benchmark run that does not pin and record the model is one deploy away from being unreproducible.

- [ ] Pin the model for the duration of the run; confirm with `aws lambda get-function-configuration ... DEFAULT_BEDROCK_MODEL_ID`
- [ ] Capture model + commit + config in the results artifact

---

## Variance: report a distribution, not a point

Two independent sources of non-determinism:

1. **Our output.** Frontier models vary run to run even at fixed temperature and pinned version.
2. **Their scoring.** `step3_judge_comments` is itself LLM-judged (`MARTIAN_MODEL`), so the same comments can score differently across runs.

A single run produces a number with an unstated error bar.

- [ ] Run the offline set **at least 3 times** per arm; publish mean and range for precision and recall
- [ ] Record the judge model used, since it is part of the measurement apparatus
- [ ] If variance is large relative to the gaps between leaderboard entries, say so plainly — that is a finding about the benchmark, and stating it is more credible than a false-precision figure

---

## Operational prerequisites

- [ ] **Create the benchmark GitHub org** and fork the 50 PRs into it per upstream instructions
- [ ] **Install the MergeWatch App** on that org
- [ ] **Billing:** `FREE_REVIEW_LIMIT` is 5 lifetime reviews per installation. 50 PRs × 3 runs × 2 arms ≈ 300 reviews. Either top up a balance or issue an OSS Program grant to the benchmark org (`scripts/grant-oss.ts`, #261) — the latter is a reasonable dogfood of that machinery, and the repos are public
- [ ] **Cost estimate:** reviews observed at roughly \$0.30–\$0.60 each at current pricing, so ≈ \$90–\$180 for the full protocol. Confirm against `us.` Bedrock rates (`packages/core/src/llm/pricing.ts`) before committing spend
- [ ] **Ensure PRs are open and non-draft**, or `skipDrafts` silently zeroes them
- [ ] Confirm reviews actually posted on all 50 before running the pipeline — a partial run scores as poor recall

---

## Publishing our own results

Independent of the upstream submission, we should host the results ourselves so we control the caveats.

- [ ] A page under `docs-site/` (suggest `overview/benchmarks.mdx`) with: both arms, mean ± range, model + commit + config + date, and an explicit statement of what the benchmark does and does not measure
- [ ] Link it from `overview/introduction.mdx`, which currently carries a competitor table with data "last verified April 2026" — a measured result is a stronger claim than a feature matrix
- [ ] **Pair it with production evidence.** We already capture per-finding outcomes on real PRs: `FindingDispositionRecord` (`surfaceCount`, `disputeCount`, `verifiedCount`, `unverifiedCount`, quiet drops) and `InstallationFPInsight` dispute rates. Aggregate dispute rate on live repositories is arguably stronger evidence than agreement with a fixed corpus, because the labels come from maintainers acting on the finding. A benchmark score plus a production dispute rate answers #242 far better than either alone
- [ ] Keep the raw run artifacts in-repo so the numbers are auditable rather than asserted

---

## Submitting upstream

- [ ] Open an issue or PR on `withmartian/code-review-benchmark` per their contributing note, adding MergeWatch to the download config with our results
- [ ] Reply on #242 with the results and a link, closing the loop with @AdrianMachado
- [ ] Offer the online benchmark too, since it resists memorization and is the better ongoing signal

---

## Honest risks worth naming up front

**We may not score well on recall**, for the reasons in the configuration section. If Arm A lands mid-table on recall and high on precision, publish it anyway and explain the tradeoff. Suppressing a result we solicited would be worse than a mediocre number, and #242 is public.

**Fixed corpora leak.** The offline 50 are public PRs from major projects and may be in training data for any model, ours included. The online benchmark exists precisely for this and is the more meaningful long-run measure.

**A single run is not evidence.** If we publish one number without variance, the first person to re-run and get something different is right to distrust everything else we claim.

## Acceptance criteria

- [ ] Both arms run ≥3× on the offline set, with mean and range published
- [ ] Model id, commit SHA, config and date recorded for every reported figure
- [ ] Results page live under `docs-site/`, linked from the introduction
- [ ] Production dispute-rate evidence published alongside the benchmark score
- [ ] Submission opened upstream at `withmartian/code-review-benchmark`
- [ ] #242 answered with a link, and closed
